### PR TITLE
feat: implement memory transaction

### PR DESCRIPTION
**Pull request by bsbds**
Depends-On: #674 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

  Implement txn API for memory engine.

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: crates/engine/src/mock_rocksdb_engine.rs [hunk 1]
```diff
@@ -197,3 +202,18 @@ impl SnapshotApi for RocksSnapshot {
         self.inner.clean().await
     }
 }
+
+/// A transaction of the `RocksEngine`
+#[derive(Copy, Clone, Debug, Default)]
+pub struct RocksTransaction;
```
Comment (Phoenix500526) [inline crates/engine/src/mock_rocksdb_engine.rs:208]: Why is this `RocksTransaction` empty?

Comment (bsbds) [reply crates/engine/src/mock_rocksdb_engine.rs:208]: It's not implement yet, I'll implement it in another PR.

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::multiple_inherent_impl)]
+
+use crate::TransactionApi;
+
+/// A transaction of the `RocksEngine`
+#[derive(Copy, Clone, Debug, Default)]
+pub struct RocksTransaction;
```
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs:8]: The derive will be removed in another PR.

[the remark this answers is not in the document]
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs:8]: see #703

**Comment by Phoenix500526** [comment]
@Mergifyio rebase

File: crates/engine/src/memory_engine/transaction.rs [hunk 3]
```diff
@@ -1,18 +1,154 @@
...
+            .ok_or_else(|| EngineError::TableNotFound(table.to_owned()))?;
+
+        if let Some(val) = state_table.get(key.as_ref()) {
+            return Ok(val.clone());
+        }
+
+        let db_inner_r = self.db.inner.read();
+        let db_table = db_inner_r
+            .get(table)
+            .ok_or_else(|| EngineError::TableNotFound(table.to_owned()))?;
```
Comment (Phoenix500526) [inline crates/engine/src/memory_engine/transaction.rs:52]: What‘s the difference between state table and db table?

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**